### PR TITLE
Added some new features and upgraded Graph API to v19.0

### PR DESCRIPTION
**Pull request by derrickobedgiu1**
This PR includes Graph API upgrade to v19.0 and some features that didn't already exist here.

For anyone looking to contribute, these are some missing features  among others:

1. Multi-Product Message Templates: [Create Template](https://developers.facebook.com/docs/whatsapp/business-management-api/message-templates/mpm-templates), [Send Template](https://developers.facebook.com/docs/whatsapp/cloud-api/guides/send-message-templates/mpm-template-messages)
2. Carousel Templates: [Create Template](https://developers.facebook.com/docs/whatsapp/business-management-api/message-templates/carousel-templates), [Send Template](https://developers.facebook.com/docs/whatsapp/cloud-api/guides/send-message-templates/carousel-template-messages) 
3. Flows: [Info](https://developers.facebook.com/docs/whatsapp/flows), [Sending](https://developers.facebook.com/docs/whatsapp/cloud-api/guides/send-messages#flows-messages)
4. [Conversational Components](https://developers.facebook.com/docs/whatsapp/cloud-api/phone-numbers/conversational-components)

Haven't had time to check if the [Template Component builder](https://github.com/netflie/whatsapp-cloud-api#send-a-template-message) is compatible with construction of 1 and 2 otherwise it might be great to add a README sample to use them.

### CODE REVIEW

File: src/Message/ButtonReply/ButtonAction.php [hunk 1]
```diff
@@ -27,4 +34,25 @@ public function buttons(): array
...
     }
+
+    /**
+     * Get the header of the button action, if it exists
+     *
+     * @return array|null The header of the button action, or null if it doesn't exist
+     */
+    public function header(): array
+    {
+        $header = [];
```
Comment (aalbarca) [inline src/Message/ButtonReply/ButtonAction.php:45]: please, if you expect that header will be always an array so it's better that you don't define $header as nullable value. You can set as default value an empty array in the constructor and remove the nullable flag.

File: src/Message/ButtonReply/ButtonAction.php [hunk 2]
```diff
@@ -27,4 +34,25 @@ public function buttons(): array
 
         return $buttonActions;
     }
+
+    /**
+     * Get the header of the button action, if it exists
+     *
+     * @return array|null The header of the button action, or null if it doesn't exist
```
Comment (aalbarca) [inline src/Message/ButtonReply/ButtonAction.php:41]: the header function don't allow return nullable values as you have declared it.

File: src/Message/ButtonReplyMessage.php [hunk 3]
```diff
@@ -8,29 +8,21 @@ class ButtonReplyMessage extends Message
...
-    private ?string $header;
-
     private string $body;
 
     private ?string $footer;
 
     private ButtonAction $action;
 
-    public function __construct(string $to, string $body, ButtonAction $action, ?string $header = null, ?string $footer = null, ?string $reply_to = null)
+    public function __construct(string $to, string $body, ButtonAction $action, ?string $footer = null, ?string $reply_to = null)
```
Comment (aalbarca) [inline src/Message/ButtonReplyMessage.php:17]: ButtonReplyMessage is not a final class, that is, someone else may have extended this class in their project.

If you change the order or remove parameters from the constructor you are breaking all projects that update the version of this package.

File: src/Message/ButtonReplyMessage.php [hunk 4]
```diff
@@ -8,29 +8,21 @@ class ButtonReplyMessage extends Message
...
     {
         $this->body = $body;
         $this->action = $action;
-        $this->header = $header;
         $this->footer = $footer;
 
         parent::__construct($to, $reply_to);
     }
 
-    public function header(): ?string
```
Comment (aalbarca) [inline src/Message/ButtonReplyMessage.php:29]: header is a public method, if you remove a public method you are breaking all projects that update the version of this package.

File: src/Message/CatalogMessage.php [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+final class CatalogMessage extends Message
+{
+    /**
+     * {@inheritdoc}
+     */
+    protected string $type = 'catalog_message';
+
+    private string $body;
+
+    private string $footer;
```
Comment (aalbarca) [inline src/Message/CatalogMessage.php:14]: $footer is defined as nullable on the constructor but no here.

File: src/Message/CatalogMessage.php [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+
+    private string $thumbnail;
```
Comment (aalbarca) [inline src/Message/CatalogMessage.php:16]: same

File: src/Message/CatalogMessage.php [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+    {
+        return $this->body;
+    }
+
+    public function footer(): ?string
+    {
+        return $this->footer;
+    }
+
+    public function productRetailerId(): ?string
```
Comment (aalbarca) [inline src/Message/CatalogMessage.php:40]: why the property and the method has different names?

File: src/Message/Media/Header.php [hunk 8]
```diff
@@ -0,0 +1,43 @@
+<?php
+
+namespace Netflie\WhatsAppCloudApi\Message\Media;
+
+class Header
```
Comment (aalbarca) [inline src/Message/Media/Header.php:5]: Final classes are always better if you are not extending

File: src/Message/MultiProduct/Action.php [hunk 9]
```diff
@@ -0,0 +1,27 @@
+<?php
+
+namespace Netflie\WhatsAppCloudApi\Message\MultiProduct;
+
+class Action
```
Comment (aalbarca) [inline src/Message/MultiProduct/Action.php:5]: Final classes are always better if you are not extending

File: src/Message/ButtonReply/ButtonCallToAction.php [hunk 10]
```diff
@@ -0,0 +1,42 @@
+<?php
+
+namespace Netflie\WhatsAppCloudApi\Message\ButtonReply;
+
+class ButtonCallToAction
```
Comment (aalbarca) [inline src/Message/ButtonReply/ButtonCallToAction.php:5]: Final classes are always better if you are not extending

File: src/Message/ButtonCallToActionMessage.php [hunk 11]
```diff
@@ -0,0 +1,92 @@
+<?php
+
+namespace Netflie\WhatsAppCloudApi\Message;
+
+use Netflie\WhatsAppCloudApi\Message\ButtonReply\ButtonCallToAction;
+
+class ButtonCallToActionMessage extends Message
```
Comment (aalbarca) [inline src/Message/ButtonCallToActionMessage.php:7]: Final classes are always better if you are not extending

File: src/Message/MultiProduct/Row.php [hunk 12]
```diff
@@ -0,0 +1,18 @@
+<?php
+
+namespace Netflie\WhatsAppCloudApi\Message\MultiProduct;
+
+class Row
```
Comment (aalbarca) [inline src/Message/MultiProduct/Row.php:5]: Final classes are always better if you are not extending

File: src/Message/MultiProduct/Section.php [hunk 13]
```diff
@@ -0,0 +1,34 @@
+<?php
+
+namespace Netflie\WhatsAppCloudApi\Message\MultiProduct;
+
+class Section
```
Comment (aalbarca) [inline src/Message/MultiProduct/Section.php:5]: Final classes are always better if you are not extending

File: src/Message/MultiProductMessage.php [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+     */
+    protected string $type = 'product_list';
+
+    private string $header;
+
+    private string $body;
+
+    private int $catalog_id;
+
+    private string $footer;
```
Comment (aalbarca) [inline src/Message/MultiProductMessage.php:20]: $footer is defined as nullable in the constructor

File: src/Message/ReactionMessage.php [hunk 15]
```diff
@@ -0,0 +1,36 @@
...
+    protected string $type = 'reaction';
+
+    private $emoji;
+
+    private $message_id;
+
+    /**
+     * {@inheritdoc}
+     */
+    public function __construct(string $to, string $message_id, string $emoji = '')
```
Comment (aalbarca) [inline src/Message/ReactionMessage.php:19]: are you sure that $emoji can to be empty string?

File: src/Message/SingleProductMessage.php [hunk 16]
```diff
@@ -0,0 +1,52 @@
...
+    /**
+     * {@inheritdoc}
+     */
+    protected string $type = 'product';
+
+    private int $catalog_id;
+
+    private string $product_retailer_id;
+
+    private string $body;
```
Comment (aalbarca) [inline src/Message/SingleProductMessage.php:16]: defined as nullable on constructor

File: src/Message/SingleProductMessage.php [hunk 17]
```diff
@@ -0,0 +1,52 @@
...
+
+    private string $footer;
```
Comment (aalbarca) [inline src/Message/SingleProductMessage.php:18]: defined as nullable on constructor

File: src/Request/MessageRequest/RequestButtonCallToActionMessage.php [hunk 18]
```diff
@@ -0,0 +1,45 @@
+<?php
+
+namespace Netflie\WhatsAppCloudApi\Request\MessageRequest;
+
+use Netflie\WhatsAppCloudApi\Request\MessageRequest;
+
+class RequestButtonCallToActionMessage extends MessageRequest
```
Comment (aalbarca) [inline src/Request/MessageRequest/RequestButtonCallToActionMessage.php:7]: Final classes are always better if you are not extending

File: src/Request/MessageRequest/RequestLocationRequestMessage.php [hunk 19]
```diff
@@ -0,0 +1,34 @@
+<?php
+
+namespace Netflie\WhatsAppCloudApi\Request\MessageRequest;
+
+use Netflie\WhatsAppCloudApi\Request\MessageRequest;
+
+class RequestLocationRequestMessage extends MessageRequest
```
Comment (aalbarca) [inline src/Request/MessageRequest/RequestLocationRequestMessage.php:7]: Final classes are always better if you are not extending

File: src/Request/MessageRequest/RequestMultiProductMessage.php [hunk 20]
```diff
@@ -0,0 +1,40 @@
+<?php
+
+namespace Netflie\WhatsAppCloudApi\Request\MessageRequest;
+
+use Netflie\WhatsAppCloudApi\Request\MessageRequest;
+
+class RequestMultiProductMessage extends MessageRequest
```
Comment (aalbarca) [inline src/Request/MessageRequest/RequestMultiProductMessage.php:7]: Final classes are always better if you are not extending

File: src/Request/MessageRequest/RequestSingleProductMessage.php [hunk 21]
```diff
@@ -0,0 +1,42 @@
+<?php
+
+namespace Netflie\WhatsAppCloudApi\Request\MessageRequest;
+
+use Netflie\WhatsAppCloudApi\Request\MessageRequest;
+
+class RequestSingleProductMessage extends MessageRequest
```
Comment (aalbarca) [inline src/Request/MessageRequest/RequestSingleProductMessage.php:7]: Final classes are always better if you are not extending

File: src/WhatsAppCloudApi.php [hunk 22]
```diff
@@ -226,14 +228,38 @@ public function sendSticker(string $to, MediaID $sticker_id): Response
...
+     * Sends a reaction to a provided message id.
+     *
+     * @param  string   $to             WhatsApp ID or phone number for the person you want to send a message to.
+     * @param  string   $emoji          The emoji to use as a reaction.
+     * @param  string   $message_id     The ID of the message to react to.
+     * @return Response
+     *
+     * @throws Response\ResponseException
+     */
+    public function sendReaction(string $to, string $message_id, string $emoji = ''): Response
```
Comment (aalbarca) [inline src/WhatsAppCloudApi.php:241]: are you sure that $emoji can to be a empty string?

File: src/WhatsAppCloudApi.php [hunk 23]
```diff
@@ -289,13 +315,23 @@ public function sendList(string $to, string $header, string $body, string $foote
...
+     *
+     * @param  string       $to             WhatsApp ID or phone number for the person you want to send a message to.
+     * @param  string       $body           The body of the message.
+     * @param  ButtonAction $action    The button action.
+     * @param  string|null  $footer      The footer text.
+     * @return Response
+     *
+     * @throws Response\ResponseException
+     */
+    public function sendButton(string $to, string $body, ButtonAction $action, ?string $footer = null): Response
```
Comment (aalbarca) [inline src/WhatsAppCloudApi.php:329]: sendButton is a public method, if you remove the arguments from a public method you are breaking all projects that update the version of this package.

**Review by aalbarca — CHANGES_REQUESTED** [review]
@derrickobedgiu1 Your work is incredible, congratulations!

File: src/Message/CatalogMessage.php (hunk 6 above)
Comment (derrickobedgiu1) [reply src/Message/CatalogMessage.php:16]: Unbundled and fixed in PR https://github.com/netflie/whatsapp-cloud-api/pull/191

File: src/Message/CatalogMessage.php (hunk 7 above)
Comment (derrickobedgiu1) [reply src/Message/CatalogMessage.php:40]: Unbundled and fixed with the appropriate method name in PR https://github.com/netflie/whatsapp-cloud-api/pull/191

File: src/WhatsAppCloudApi.php (hunk 22 above)
Comment (derrickobedgiu1) [reply src/WhatsAppCloudApi.php:241]: Oh yes! I mentioned it in PR https://github.com/netflie/whatsapp-cloud-api/pull/190

File: src/WhatsAppCloudApi.php (hunk 23 above)
Comment (derrickobedgiu1) [reply src/WhatsAppCloudApi.php:329]: Change Reverted

File: src/Message/ButtonReply/ButtonAction.php (hunk 1 above)
Comment (derrickobedgiu1) [reply src/Message/ButtonReply/ButtonAction.php:45]: Change reverted. It was to support media header for button reply but was breaking

File: src/Message/ButtonReply/ButtonAction.php (hunk 2 above)
Comment (derrickobedgiu1) [reply src/Message/ButtonReply/ButtonAction.php:41]: Same with above.

File: src/Message/ButtonReplyMessage.php (hunk 3 above)
Comment (derrickobedgiu1) [reply src/Message/ButtonReplyMessage.php:17]: Reverted and not included in unbundled PRs

File: src/Message/ButtonReplyMessage.php (hunk 4 above)
Comment (derrickobedgiu1) [reply src/Message/ButtonReplyMessage.php:29]: Same as reply given before

**Pull request closed without merging**